Source/WebCore/Modules/mediastream/VideoTrackGenerator.cpp

```diff
@@ -27,24 +27,166 @@
 
 #if ENABLE(MEDIA_STREAM)
 
+#include "Exception.h"
+#include "JSWebCodecsVideoFrame.h"
 #include "MediaStreamTrack.h"
 #include "WritableStream.h"
+#include "WritableStreamSink.h"
 
 namespace WebCore {
 
 WTF_MAKE_ISO_ALLOCATED_IMPL(VideoTrackGenerator);
 
-VideoTrackGenerator::VideoTrackGenerator(ScriptExecutionContext&)
+ExceptionOr<Ref<VideoTrackGenerator>> VideoTrackGenerator::create(ScriptExecutionContext& context)
 {
+    auto source = Source::create();
+    auto sink = Sink::create(Ref { source });
+    auto writableOrException = WritableStream::create(*JSC::jsCast<JSDOMGlobalObject*>(context.globalObject()), Ref { sink });
+
+    if (writableOrException.hasException())
+        return writableOrException.releaseException();
+
+    // FIXME: Maybe we should have the writable buffer frames until the source is actually started.
+    callOnMainThread([source] {
+        source->start();
+    });
+
+    auto track = MediaStreamTrack::create(context, MediaStreamTrackPrivate::create(Logger::create(&context), WTFMove(source), [identifier = context.identifier()](Function<void()>&& task) {
```

Comment:
The newly created Logger won't respect the document's ephemeral state.

Reply:
I'll add logger->setEnabled(sessionID().isAlwaysOnLoggingAllowed());